audit/generic_auditor.py

```diff
@@ -0,0 +1,652 @@
+                    if not reads:
+                        logging.info(
+                            f'Sample {sample_internal_id} has no sequence reads'
+                        )
+                        continue
+                    # support up to three levels of nesting, because:
+                    #
+                    #   - reads is typically an array
+                    #       - handle the case where it's not an array
+                    #   - fastqs exist in pairs.
+                    # eg:
+                    #   - reads: cram   # this shouldn't happen, but handle it anyway
+                    #   - reads: [cram1, cram2]
+                    #   - reads: [[fq1_forward, fq1_back], [fq2_forward, fq2_back]]
+                    if isinstance(reads, dict):
+                        sequence_filepaths_filesizes[sequence.get('id')].append(
+                            (reads.get('location'), reads.get('size'))
+                        )
+                        continue
+                    if not isinstance(reads, list):
+                        logging.error(f'Invalid read type: {type(reads)}: {reads}')
+                        continue
+                    for read in reads:
+                        if isinstance(read, list):
+                            for inner_read in read:
+                                if not isinstance(inner_read, dict):
+                                    logging.error(
+                                        f'Got {type(inner_read)} read, expected dict: {inner_read}'
+                                    )
+                                    continue
+                                sequence_filepaths_filesizes[sequence.get('id')].append(
+                                    (inner_read.get('location'), inner_read.get('size'))
+                                )
+                        else:
+                            if not isinstance(read, dict):
+                                logging.error(
+                                    f'Got {type(inner_read)} read, expected dict: {inner_read}'
+                                )
+                                continue
+                            sequence_filepaths_filesizes[sequence.get('id')].append(
+                                (read.get('location'), read.get('size'))
+                            )
+
+        return seq_id_sample_id_map, sequence_filepaths_filesizes
+
+    def get_sequences_from_participants(
+        self, participants: list[dict]
+    ) -> tuple[dict[Any, Any], defaultdict[Any, list[tuple[Any, Any]]]]:
+        """Return latest sequence ids for internal sample ids"""
+        all_sequences = [
+            {
+                sample.get('id'): sample.get('sequences')
+                for sample in participant.get('samples')
+            }
+            for participant in participants
+        ]
+
+        return self.get_sequence_mapping(all_sequences)
```

Comment:
Can these methods be merged or separated? You're making yourself a unit testing headache if you tie two methods together